### PR TITLE
Scale color output in the mobile renderer to provide HDR support

### PR DESCRIPTION
**Pull request by BastiaanOlij**
As we are now using an A2R10G10B10 UNORM buffer our color values can only be between 0.0 and 1.0 disabling full HDR support on the mobile renderer. 

By scaling the color output we can support a 0.0 to 2.0 range. This involves dividing by 2.0 at the output of our scene renderer, and multiplying by 2.0 in our tonemapper. Similar changes are needed in some other parts of the engine such as reading from screen_texture and reading from reflection maps both in reflection probes and for our sky.

Because the clustered renderer and mobile renderer share some code (mostly the lighting code) the `color_scale` input has been added to the clustered `scene_data` as well but remains large unused and should remain at its default `1.0`.

### CODE REVIEW

File: servers/rendering/renderer_rd/effects_rd.cpp
Comment (BastiaanOlij) [inline servers/rendering/renderer_rd/effects_rd.cpp:63]: So these are the regression fixesses. As the `luminance_reduce` shaders are now unused and not initialised on the mobile renderer, when we abuse them in this way to setup uniform sets for other shaders, stuff break... 

As discussed with Reduz, I plan to split all the effects into their own classes to prevent stuff like this.

File: servers/rendering/renderer_rd/effects_rd.cpp
Comment (BastiaanOlij) [inline servers/rendering/renderer_rd/effects_rd.cpp:109]: Ditto here

**Comment by BastiaanOlij** [comment]
This is on hold until I finish #50942 in which I'm also addressing the above breakages in the mobile renderer.

**Comment by reduz** [comment]
regarding reading the textures and doing the right adjustment, you can add the code here:
https://github.com/godotengine/godot/blob/master/servers/rendering/renderer_rd/shader_compiler_rd.cpp#L1162
* Add an optional mode (likely a constant that by default is 1.0) called screen_texture_multiply or something you can set.
* Detect that this is reading from a texture (already done in *is_texture_func* )
* Detect that the texture name is SCREEN_TEXTURE ( think you can do this by checking *texture_uniform* ?)
* If all this matches, you need to wrap the whole thing in parenthesis () and after the end where it does `code += ")"`` add something like: `) *  "+rtos(screen_texture_multiply)+")"`

File: servers/rendering/renderer_rd/effects_rd.cpp
Comment (BastiaanOlij) [reply servers/rendering/renderer_rd/effects_rd.cpp:63]: This has now been dealt with by rewriting these shaders to fragment shaders in another PR that was already merged.

**Comment by BastiaanOlij** [comment]
> 
> 
> regarding reading the textures and doing the right adjustment, you can add the code here:
> https://github.com/godotengine/godot/blob/master/servers/rendering/renderer_rd/shader_compiler_rd.cpp#L1162
> 
>     * Add an optional mode (likely a constant that by default is 1.0) called screen_texture_multiply or something you can set.
> 
>     * Detect that this is reading from a texture (already done in _is_texture_func_ )
> 
>     * Detect that the texture name is SCREEN_TEXTURE ( think you can do this by checking _texture_uniform_ ?)
> 
>     * If all this matches, you need to wrap the whole thing in parenthesis () and after the end where it does `code += ")"` add something like: `) *  "+rtos(screen_texture_multiply)+")"`

@reduz Seeing I've already added the scale factor to the scene_data I've just added that in. It was already available as I'm using it in several other places of the shader.
Other then that it seems to be creating the right shader code, I can't test it because reading from screen texture was already broken. Seeing we don't ever seem to make a copy of the render buffer to use as a texture to read to I wonder if this is the bit that is missing here. But that is for another PR methinks.

Let me know if you're happy with the approach so far or if you want it to be changed.

File: servers/rendering/renderer_rd/effects_rd.cpp
Comment (BastiaanOlij) [inline servers/rendering/renderer_rd/effects_rd.cpp:909]: Forgot about this, this value should be a parameter/setting

File: servers/rendering/renderer_rd/effects_rd.h
Comment (reduz) [inline servers/rendering/renderer_rd/effects_rd.h:785]: does it make sense to scale the luminance reduction? I would just multiply by scale in the tonemapper, so no precision is lost.

File: servers/rendering/renderer_rd/forward_clustered/render_forward_clustered.h
Comment (reduz) [inline servers/rendering/renderer_rd/forward_clustered/render_forward_clustered.h:291]: This is not really needed in clustered.

File: servers/rendering/renderer_rd/forward_mobile/render_forward_mobile.cpp
Comment (reduz) [inline servers/rendering/renderer_rd/forward_mobile/render_forward_mobile.cpp:274]: I have the feeling this should be called luminance_scale rather than color_scale

Comment (reduz) [reply servers/rendering/renderer_rd/forward_mobile/render_forward_mobile.cpp:274]: or luminance_multiplier?

File: servers/rendering/renderer_rd/shader_compiler_rd.cpp
Comment (reduz) [inline servers/rendering/renderer_rd/shader_compiler_rd.cpp:1201]: All this mechanism most likely needs some enabling in ShaderCompilerRD because, as an example, in 2D or in the clustered rendering, you dont want to use it.

File: servers/rendering/renderer_rd/shaders/scene_forward_lights_inc.glsl
Comment (reduz) [inline servers/rendering/renderer_rd/shaders/scene_forward_lights_inc.glsl:947]: I would probably make this a specialization constant in the mobile renderer and a regular constant (1.0) in the clustered one.

Comment (reduz) [reply servers/rendering/renderer_rd/shaders/scene_forward_lights_inc.glsl:947]: See #51635 for reference (which I guess should be merged first).

Comment (reduz) [reply servers/rendering/renderer_rd/shaders/scene_forward_lights_inc.glsl:947]: after merging that PR, on line 400 of shader_forward_mobile.glsl you can add:

`layout(constant_id = 15) const float sc_luminance_multiplier = 2.0;`

This has two advantages. First is that you can also use it to divide at the end of the shader (and it will be turned to a multiplication by the compiler). Second, that in the clustered one you can simply do something like:

`const float sc_luminance_multiplier = 1.0;`

and this will be optimized away.

**Review by reduz — CHANGES_REQUESTED** [review]

File: servers/rendering/renderer_rd/effects_rd.h
Comment (BastiaanOlij) [reply servers/rendering/renderer_rd/effects_rd.h:785]: We render into a single channel buffer here so there is alot more precision then the source color. I guess we could do the multiplication much later

File: servers/rendering/renderer_rd/forward_clustered/render_forward_clustered.h
Comment (BastiaanOlij) [reply servers/rendering/renderer_rd/forward_clustered/render_forward_clustered.h:291]: The problem is that the clustered and mobile renderer share code, one needs it, the other doesn't...

File: servers/rendering/renderer_rd/forward_mobile/render_forward_mobile.cpp
Comment (BastiaanOlij) [reply servers/rendering/renderer_rd/forward_mobile/render_forward_mobile.cpp:274]: I'm open for either

File: servers/rendering/renderer_rd/shader_compiler_rd.cpp
Comment (BastiaanOlij) [reply servers/rendering/renderer_rd/shader_compiler_rd.cpp:1201]: Yeah that makes sense, I'm still learning the parser code so not sure what would be the easiest way to switch on it? Can I check if this is a spatial shader?

File: servers/rendering/renderer_rd/shaders/scene_forward_lights_inc.glsl
Comment (BastiaanOlij) [reply servers/rendering/renderer_rd/shaders/scene_forward_lights_inc.glsl:947]: Ok I'll look into that, need to learn how that works :)

File: servers/rendering/renderer_rd/forward_clustered/render_forward_clustered.h
Comment (BastiaanOlij) [reply servers/rendering/renderer_rd/forward_clustered/render_forward_clustered.h:291]: Now solved with a constant

File: servers/rendering/renderer_rd/forward_mobile/render_forward_mobile.cpp
Comment (BastiaanOlij) [reply servers/rendering/renderer_rd/forward_mobile/render_forward_mobile.cpp:274]: Settled on `luminance_multiplier`

File: servers/rendering/renderer_rd/effects_rd.h
Comment (BastiaanOlij) [reply servers/rendering/renderer_rd/effects_rd.h:785]: I've moved this into the tonemapper, there isn't any difference performance wise but doing this in one place instead of two makes sense.

File: servers/rendering/renderer_rd/shader_compiler_rd.cpp
Comment (BastiaanOlij) [reply servers/rendering/renderer_rd/shader_compiler_rd.cpp:1201]: @reduz I added a `apply_luminance_multiplier` boolean to `DefaultIdentifierActions` that is only set by the shader compiler for the mobile renderer. Not sure if this is what you had in mind but it seems to do the trick :)

**Review by reduz — APPROVED** [review]

**Pull request merged**
